### PR TITLE
Language export import

### PR DESCRIPTION
**Pull request by CarlosNZ**
Requires [back-end PR #748](https://github.com/openmsupply/application-manager-server/pull/748)

Screen Shot 2022-05-08 at 6 43 59 PM

Provides UI for `/admin/localisation` page, with following functionality:

- Enable/disable installed languages
- Export languages to CSV (for editing in Google Sheets, for example)
- Import languages from CSV

Needs `yarn install` to install `papa-parse` -- CSV [parsing library](https://www.papaparse.com/)

### Recommended usage

- Use this Google sheet to manage: https://docs.google.com/spreadsheets/d/1RE80Rqm5_TI3k1ibm9C5gEyGjk9l91Od4244D6gZqCw/edit#gid=725396254
- Export as CSV from page
- In Google Sheet, "Import" CSV:
  - Select "Replace current sheet" for "import location"
  - UN-check "Convert text to numbers, dates, and formulas"
- Import and data will be update. Browse, edit accordingly
- Export by "File -> Download -> CSV"
- Re-import using UI!
Screen Shot 2022-05-08 at 6 36 51 PM

I have some more thoughts to be able to also include plugin strings (and eventually localised template strings) so all translation management can be done in one place, but this is a first iteration.

(Personally I think editing all strings in a single spreadsheet is a _much_ better way to manage it than making our own UI with editing strings one by one, which would be cumbersome and not as good -- but open to discussion if others think otherwise)

### CODE REVIEW

File: src/components/Main/ModalWarning.tsx [hunk 1]
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import { Button, Header, Icon, Modal, ModalProps } from 'semantic-ui-react'
 
 const ModalWarning: React.FC<ModalProps> = (showModal) => {
-  const { title, message, option, onClick, ...modalProps } = showModal
+  const { title, message, option, optionCancel, onClick, onCancel, ...modalProps } = showModal
```
Comment (CarlosNZ) [inline src/components/Main/ModalWarning.tsx:5]: Minor addition to Modal options, surprised we didn't already have a "Cancel" button on these Modals. I think we should move to a global Modal UI that works the same was as the [ToastProvider](#1189), but that's for another day.

File: src/containers/User/UserArea.tsx [hunk 2]
```diff
@@ -102,9 +102,9 @@ const MainMenuBar: React.FC<MainMenuBarProps> = ({ templates, outcomes }) => {
       text: strings.MENU_ITEM_ADMIN_LOOKUP_TABLES,
       value: '/admin/lookup-tables',
     },
-    { key: 'outcomes', text: strings.MENU_ITEM_ADMIN_OUTCOME_CONFIG, value: '/admin/outcomes' },
-    { key: 'permissions', text: strings.MENU_ITEM_ADMIN_PERMISSIONS, value: '/admin/permissions' },
-    { key: 'plugins', text: strings.MENU_ITEM_ADMIN_PLUGINS, value: '/admin/plugins' },
+    // { key: 'outcomes', text: strings.MENU_ITEM_ADMIN_OUTCOME_CONFIG, value: '/admin/outcomes' },
+    // { key: 'permissions', text: strings.MENU_ITEM_ADMIN_PERMISSIONS, value: '/admin/permissions' },
+    // { key: 'plugins', text: strings.MENU_ITEM_ADMIN_PLUGINS, value: '/admin/plugins' },
```
Comment (CarlosNZ) [inline src/containers/User/UserArea.tsx:107]: Disabling these menu items until we actually have something to show for them.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "application-manager-web-app",
-  "version": "0.2.0-19",
+  "version": "0.2.0-23",
```
Comment (nmadruga) [inline package.json:3]: Hmmm 
This might be hard when we are merging everything, but not a big deal...

File: src/components/Admin/AdminLocalisations.tsx [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+  const [showModalWarning, setShowModalWarning] = useState<ModalProps>({ open: false })
+  const [hoverIndex, setHoverIndex] = useState<number | null>(null)
+
+  const handleSelect = async (language: LanguageOption, index: number) => {
+    const enabled = !language.enabled
+    const result = await postRequest({
+      url: `${config.serverREST}/admin/enable-language?code=${language.code}&enabled=${enabled}`,
+    })
+    if (result.success) {
+      console.log('Language updated')
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:28]: Useful to add to which code here - or just remove the log

File: src/components/Admin/AdminLocalisations.tsx [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+      option: strings.BUTTON_CONFIRM,
+      optionCancel: strings.OPTION_CANCEL,
+      onCancel: () => setShowModalWarning({ open: false }),
+      onClick: async () => {
+        setShowModalWarning({ open: false })
+        const result = await postRequest({
+          url: `${config.serverREST}/admin/remove-language?code=${language.code}`,
+        })
+        if (result.success) {
+          console.log('Language removed')
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:53]: Same

File: src/components/Admin/AdminLocalisations.tsx [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+  const handleExport = async () => {
+    const result = await exportLanguages(
+      languageOptions.filter((lang) => exportDisabled || lang.enabled)
+    )
+    if (result.success)
+      showToast({
+        title: 'Languages exported',
+        text: 'File has been downloaded',
+        style: 'success',
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:84]: As mentioned in the Toast PR, should we also localise this considering local Admin staff working in another language? Or that's too much? I just consider that if we have localised things, maybe it's best to keep doing it for new things (and check what needs to be included - hint: Template Builder)

File: src/components/Admin/AdminLocalisations.tsx [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+      })
+    else {
+      showToast({
+        title: strings.LOCALISATION_REMOVE_ERROR,
+        text: `${strings.LOCALISATION_EXPORT_PROBLEM}: ${result?.message}`,
+        style: 'error',
+      })
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:91]: Oh, you have actually done this. Maybe just missing for the one before : )

File: src/components/Admin/AdminLocalisations.tsx [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+      console.error(result.message)
+    }
+  }
+
+  const handleFileImport = async (e: any) => {
+    const file = e.target.files[0]
+    const reader = new FileReader()
+    reader.onload = (event) => {
+      if (!event.target?.result) {
+        showToast({ title: 'Error', text: strings.LOCALISATION_FILE_PROBLEM, style: 'error' })
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:101]: Localise: `Error`

File: src/components/Admin/AdminLocalisations.tsx [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+      importLanguages(event.target.result as string, importDisabled, strings).then(
+        ({ success, message }) => {
+          if (success) {
+            showToast({
+              title: strings.LOCALISATION_INSTALL_SUCCESS,
+              text: message,
+              style: 'success',
+            })
+          } else {
+            showToast({ title: 'Error', text: message, style: 'error' })
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:113]: Here too. Localise `Error`

File: src/components/Admin/AdminLocalisations.tsx [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      )
+    }
+    reader.readAsText(file)
+  }
+
+  return (
+    <div id="localisation-panel">
+      <ModalWarning {...showModalWarning} />
+      <Header as="h1">Localisation</Header>
+      <Header as="h4">Currently installed languages</Header>
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:126]: Localise...

File: src/components/Admin/AdminLocalisations.tsx [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    >
+      <Checkbox checked={language.enabled} onChange={() => handleSelect(language, index)} />
+      <div className="flex-column">
+        <p style={{ marginBottom: 0 }}>
+          <strong>{language.languageName}</strong>
+        </p>
+        <p className="smaller-text">{language.description}</p>
+      </div>
+      <p style={{ fontSize: '2em', flexGrow: 1, textAlign: 'right' }}>{language.flag}</p>
+    </div>
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:209]: Would be feasible to move some of the styles to LESS files?

File: src/components/Admin/AdminLocalisations.tsx [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        title: strings.LOCALISATION_REMOVE_ERROR,
+        text: `${strings.LOCALISATION_EXPORT_PROBLEM}: ${result?.message}`,
+        style: 'error',
+      })
+      console.error(result.message)
+    }
+  }
+
+  const handleFileImport = async (e: any) => {
+    const file = e.target.files[0]
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:97]: Should we flag if more than one is attempted? Or not important?

File: src/components/Main/ModalWarning.tsx (hunk 1 above)
Comment (nmadruga) [reply src/components/Main/ModalWarning.tsx:5]: The cancel used to be "click outside"

File: src/containers/Review/ReviewSubmit.tsx [hunk 13]
```diff
@@ -161,7 +161,9 @@ const ReviewSubmitButton: React.FC<ReviewSubmitProps & ReviewSubmitButtonProps>
   }
 
   const onClick = async () => {
-    const firstIncompleteReviewPage = structure.firstIncompleteReviewPage
+    const firstIncompleteReviewPage = assignment.isFinalDecision
+      ? null
+      : structure.firstIncompleteReviewPage
```
Comment (nmadruga) [inline src/containers/Review/ReviewSubmit.tsx:166]: Hmmm seems like a slip from different PR?

File: src/containers/Review/assignment/AssignmentRows.tsx [hunk 14]
```diff
@@ -14,38 +16,54 @@ interface AssignmentRowsProps {
...
+                  setAssignedSections={(assignedSections) =>
+                    setAssignedSectionsByLevel({
+                      ...assignedSectionsByLevel,
+                      [level]: assignedSections,
+                    })
+                  }
+                  setEnableSubmit={setEnableSubmit}
+                  setAssignmentError={setAssignmentError}
+                />
+              )}
```
Comment (nmadruga) [inline src/containers/Review/assignment/AssignmentRows.tsx:66]: Can you merge `develop` into this? I think it will remove some of the stuff that is unrelated

**Review by nmadruga — COMMENTED** [review]
This is a very cool feature. Thanks Carl! I'm trying to sort out - how to account for the work done here. Which we need to have a simpler way to do, any way it was a very good addition!

Things I'm very happy about:

- List of diff on KEYS showing on translated file but not on original language
Screen Shot 2022-05-24 at 10 33 13 AM


TO be continued...

**Comment by nmadruga** [comment]
Will also continue the review of this one tomorrow. 

Pending some merging so the code get's cleared from other PRs

File: src/components/Main/ModalWarning.tsx (hunk 1 above)
Comment (CarlosNZ) [reply src/components/Main/ModalWarning.tsx:5]: Yeah, and it still is. I just thought it was potentially confusing to not have an explicit "Cancel" button too. It's pretty standard for a confirmation dialog.

File: package.json (hunk 3 above)
Comment (CarlosNZ) [reply package.json:3]: That's fine, cos the version number that ends up in develop will be the version number of the last build that was done (from any branch) one all PRs are merged.

File: src/components/Admin/AdminLocalisations.tsx (hunk 4 above)
Comment (CarlosNZ) [reply src/components/Admin/AdminLocalisations.tsx]: Yup, have added code: a5a9d8f9

File: src/components/Admin/AdminLocalisations.tsx (hunk 7 above)
Comment (CarlosNZ) [reply src/components/Admin/AdminLocalisations.tsx:91]: Ah, yes -- good spotting! I even made the strings in `defaultLanguageStrings`, but forgot to apply them here. 

a5a9d8f9

File: src/components/Admin/AdminLocalisations.tsx (hunk 11 above)
Comment (CarlosNZ) [reply src/components/Admin/AdminLocalisations.tsx]: Yeah, okay, have moved the general "row" style to LESS file. Left the very specific flag formatter inline.
4a317d3b

File: src/components/Admin/AdminLocalisations.tsx (hunk 12 above)
Comment (CarlosNZ) [reply src/components/Admin/AdminLocalisations.tsx:97]: Already has `multiple={false}` in the input definition ;)

File: src/containers/Review/ReviewSubmit.tsx (hunk 13 above)
Comment (CarlosNZ) [reply src/containers/Review/ReviewSubmit.tsx]: Okay, it should be all good now.

File: src/containers/Review/assignment/AssignmentRows.tsx (hunk 14 above)
Comment (CarlosNZ) [reply src/containers/Review/assignment/AssignmentRows.tsx]: Okay, it should be all good now.

File: src/components/Admin/AdminLocalisations.tsx [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  const handleSelect = async (language: LanguageOption, index: number) => {
+    const enabled = !language.enabled
+    const result = await postRequest({
+      url: `${config.serverREST}/admin/enable-language?code=${language.code}&enabled=${enabled}`,
+    })
+    if (result.success) {
+      console.log(`Language updated: ${language.code}`)
+      refetchLanguages()
+    } else {
+      showToast({ title: 'Error', text: result.message, style: 'error' })
```
Comment (nmadruga) [inline src/components/Admin/AdminLocalisations.tsx:31]: Still missing translation for string `Error`

File: src/contexts/Localisation.tsx [hunk 16]
```diff
@@ -118,21 +125,53 @@ export function LanguageProvider({
...
   }
 
+  // If the selected language is no longer valid (either disabled or uninstalled), then try default, or fallback to first available
+  const getValidLanguageCode = () => {
+    if (languageOptions.some((lang) => lang.enabled && lang.code === selectedLanguageCode))
+      return selectedLanguageCode
+    if (languageOptions.some((lang) => lang.enabled && lang.code === defaultLanguageCode))
+      return defaultLanguageCode
+    else {
+      console.log('Invalid language code, falling back to first available')
```
Comment (nmadruga) [inline src/contexts/Localisation.tsx:135]: Maybe good to add which is the first available code?

File: src/utils/localisation/exportLanguages.ts [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  const row1Name = [
+    'Language:',
+    'System Default',
+    ...languageOptions.map((opt) => opt.languageName),
+  ]
+  const row2Description = [
+    'Description:',
+    'Internal values from which all others are translated',
+    ...languageOptions.map((opt) => opt.description),
+  ]
```
Comment (nmadruga) [inline src/utils/localisation/exportLanguages.ts:51]: Should this be localised too?

File: src/utils/localisation/exportLanguages.ts [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    row1Name,
+    row2Description,
+    row3code,
+    row4Flag,
+    row5Enabled,
+    [],
+    ['KEYS'],
+    ...translationRows,
+    [],
+    ['__ORPHANS__', 'Translations found in language files that are no longer in use'],
```
Comment (nmadruga) [inline src/utils/localisation/exportLanguages.ts:74]: And this?

**Review by nmadruga — CHANGES_REQUESTED** [review]
Ok - Looks great. But I though I would find the same explanation from the PR on one doc in the Front-end. Do you mind adding one with the same description from there please?

File: src/components/Admin/AdminLocalisations.tsx (hunk 15 above)
Comment (CarlosNZ) [reply src/components/Admin/AdminLocalisations.tsx]: Oops -- thanks 0ebf11af

File: src/contexts/Localisation.tsx (hunk 16 above)
Comment (CarlosNZ) [reply src/contexts/Localisation.tsx:135]: Sure: 96a1c966

File: src/utils/localisation/exportLanguages.ts (hunk 17 above)
Comment (CarlosNZ) [reply src/utils/localisation/exportLanguages.ts:51]: I dunno -- these go onto the CSV output and I don't know if we want to worry about making the spreadsheet localised just yet. Especially since anyone who uses it would probably need to be an English speaker to make sense of the default strings.

I was kind of imagining that we might have a shared Google sheet for all the different languages which people can keep updated with their own language, in which case this wouldn't be useful to localise. But that's a discussion fo another day -- can we leave these for now?

**Comment by CarlosNZ** [comment]
Just a reminder to myself to do before merging this: the "EXPORT" doesn't show any error if it fails -- it should show a Toast with the error message

**Comment by CarlosNZ** [comment]
> But I though I would find the same explanation from the PR on one doc in the Front-end. Do you mind adding one with the same description from there please?

Done! 1cf5a478

File: src/utils/localisation/exportLanguages.ts (hunk 17 above)
Comment (nmadruga) [reply src/utils/localisation/exportLanguages.ts]: Yep. No problem. I liked you idea of the spreadsheet with each localisation, although I understood as each project would be using theirs. Also because the sheet is currently storing all the languages that are part of a project.

**Review by nmadruga — APPROVED** [review]
Nice. Thanks for the changes : )
